Update coherence and orphan rules documentation to match RFC 2451 (#703)
rust-lang/reference: src/items/implementations.md

### Trait Implementation Coherence

A trait implementation is considered incoherent if either the orphan rules check fails
or there are overlapping implementation instances.

Two trait implementations overlap when there is a non-empty intersection of the
traits the implementation is for, the implementations can be instantiated with
the same type. <!-- This is probably wrong? Source: No two implementations can
be instantiable with the same set of types for the input type parameters. -->

#### Orphan rules

> Given `impl<P1..=Pn> Trait<T1..=Tn> for T0`, an impl is valid only if at

Reviewer: ```suggestion
Given `impl<P1..=Pn> Trait<T1..=Tn> for T0`, an `impl` is valid only if at
```

least one of the following is true:

- `Trait` is a [local trait]
- All of
  - At least one of the types `T0..=Tn` must be a [local type]. Let `Ti` be the
    first such type.
  - No [uncovered type] parameters `P1..=Pn` may appear in `T0..Ti` (excluding
    `Ti`)

> We only restrict the appearance of *uncovered* type parameters. Once again, it is

Reviewer: ```suggestion
Only the appearance of *uncovered* type parameters is restricted.
```

> important to note that for the purposes of coherence, [fundamental types] are

Reviewer: ```suggestion
Note that for the purposes of coherence, [fundamental types] are
```
